### PR TITLE
[ISSUE-367] Daemonset selection based on kernel version

### PR DESCRIPTION
**Pull request by safronovD**
### Issue dell/csi-baremetal#367

Implement https://github.com/dell/csi-baremetal/pull/368:

- update sigs.k8s.io/controller-runtime -> 0.7.2, k8s.io/client-go -> 0.19.2
- add context to each controller from main
- move kernel version parser from csi-baremetal-node-controller
- watch k8sNode, reconcile CSI Deployment on create, update kernelVersion, update labels events
- add list of supported platform
- create specific daemonset for each platform if it exists in cluster
- add unit tests for daemonset selection logic (using "k8s.io/client-go/kubernetes/fake")
- update pr.yaml (add testing stage)
- update go to 1.16 (we use it in pr validation)

## PR checklist
- [x] Add link to the issue
- [x] Choose PR label
- [x] New unit tests added
- [x] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing
```
user@user-VirtualBox ~/w/s/g/d/csi-baremetal> kubectl get pod -o wide 
NAME                                             READY   STATUS    RESTARTS   AGE     IP                NODE               NOMINATED NODE   READINESS GATES
controller-manager-6897895998-4fr5m              1/1     Running   0          8m29s   192.168.7.197     csi-ubuntu20-101   <none>           <none>
csi-baremetal-controller-7bf984f984-s78fb        4/4     Running   0          7m45s   192.168.147.141   csi-centos8-103    <none>           <none>
csi-baremetal-node-2mmsl                         4/4     Running   0          7m46s   192.168.2.206     csi-centos7-102    <none>           <none>
csi-baremetal-node-controller-65c7f99b45-wqmlc   1/1     Running   0          7m45s   192.168.2.207     csi-centos7-102    <none>           <none>
csi-baremetal-node-dq9fh                         4/4     Running   0          7m46s   192.168.21.141    csi-ubuntu18-100   <none>           <none>
csi-baremetal-node-kernel-5.4-mpwbx              4/4     Running   0          7m45s   192.168.7.198     csi-ubuntu20-101   <none>           <none>
csi-baremetal-node-rwl6k                         4/4     Running   0          7m46s   192.168.147.140   csi-centos8-103    <none>           <none>
csi-baremetal-se-4kmfs                           1/1     Running   0          7m45s   10.52.203.100     csi-ubuntu18-100   <none>           <none>
csi-baremetal-se-patcher-q8g4p                   1/1     Running   0          7m45s   192.168.21.142    csi-ubuntu18-100   <none>           <none>
web-0                                            1/1     Running   0          2m49s   192.168.7.199     csi-ubuntu20-101   <none>           <none>
web-1                                            1/1     Running   0          2m49s   192.168.2.208     csi-centos7-102    <none>           <none>
web-2                                            1/1     Running   0          2m48s   192.168.147.142   csi-centos8-103    <none>           <none>
web-3                                            1/1     Running   0          2m48s   192.168.21.143    csi-ubuntu18-100   <none>           <none>
```

### CODE REVIEW

File: controllers/deployment_controller.go [hunk 1]
```diff
@@ -106,23 +114,108 @@ func (r *DeploymentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
...
+					Name:      dep.Name,
+					Namespace: dep.Namespace,
+				}})
+		}
+
+		return requests
+	}),
+		predicate.Funcs{
+			CreateFunc: func(createEvent event.CreateEvent) bool {
+				return true
```
Comment (AnastasiaShemyakinskaya) [inline controllers/deployment_controller.go:171]: Does this filter work by default?

File: pkg/scheduler_patcher.go [hunk 2]
```diff
@@ -1,6 +1,9 @@
 package pkg
 
 import (
+	"context"
+	"github.com/dell/csi-baremetal-operator/pkg/common"
```
Comment (AnastasiaShemyakinskaya) [inline pkg/scheduler_patcher.go:5]: Import order

File: pkg/scheduler_extender.go [hunk 3]
```diff
@@ -1,6 +1,8 @@
 package pkg
 
 import (
+	"context"
+	"github.com/dell/csi-baremetal-operator/pkg/common"
```
Comment (AnastasiaShemyakinskaya) [inline pkg/scheduler_extender.go:5]: Import order

File: controllers/deployment_controller.go (hunk 1 above)
Comment (safronovD) [reply controllers/deployment_controller.go]: Yes, you're right. I expected only events, which we set here, would work. I deleted this filter and changed only UpdateFunc

File: pkg/scheduler_patcher.go (hunk 2 above)
Comment (safronovD) [reply pkg/scheduler_patcher.go]: fixed

File: pkg/scheduler_extender.go (hunk 3 above)
Comment (safronovD) [reply pkg/scheduler_extender.go]: fixed

File: pkg/node/node.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	return "default"
+}
+
+// Set is needed to check if one type of platform is exists in current cluster
+type Set map[string]bool
+
+// createNeedToDeploySet returns set of platform-names
+func createNeedToDeploySet() Set {
```
[the remark this answers is not in the document]
Comment (safronovD) [reply pkg/node/node.go]: renamed `createPlatformsSet`

**Review by yurkov-anton — APPROVED** [review]

File: controllers/deployment_controller.go [hunk 5]
```diff
@@ -51,9 +57,9 @@ const (
 // +kubebuilder:rbac:groups=csi-baremetal.dell.com,resources=deployments,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=csi-baremetal.dell.com,resources=deployments/status,verbs=get;update;patch
 
-func (r *DeploymentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	ctx := context.Background()
+func (r *DeploymentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	log := r.Log.WithValues("deployment", req.NamespacedName)
+	log.Info("reconcile!!!")
```
Comment (mishoyama) [inline controllers/deployment_controller.go:62]: Seems to be debug message. Please remove.

File: go.sum [hunk 6]
```diff
@@ -374,85 +663,178 @@ golang.org/x/tools v0.0.0-20190226205152-f727befe758c/go.mod h1:9Yl7xja0Znq3iFh3
...
+sigs.k8s.io/controller-runtime v0.7.2/go.mod h1:pJ3YBrJiAqMAZKi6UVGuE98ZrroV1p+pIhoHsMm9wdU=
+sigs.k8s.io/structured-merge-diff v0.0.0-20190525122527-15d366b2352e h1:4Z09Hglb792X0kfOBBJUPFEyvVfQWrYT/l8h5EKA6JQ=
 sigs.k8s.io/structured-merge-diff v0.0.0-20190525122527-15d366b2352e/go.mod h1:wWxsB5ozmmv/SG7nM11ayaAW51xMvak/t1r0CSlcokI=
-sigs.k8s.io/structured-merge-diff v1.0.1-0.20191108220359-b1b620dd3f06/go.mod h1:/ULNhyfzRopfcjskuui0cTITekDduZ7ycKN3oUT9R18=
-sigs.k8s.io/yaml v1.1.0 h1:4A07+ZFc2wgJwo8YNlQpr1rVlgUDlxXHhPJciaPY5gs=
+sigs.k8s.io/structured-merge-diff/v4 v4.0.1 h1:YXTMot5Qz/X1iBRJhAt+vI+HVttY0WkSqqhKxQ0xVbA=
+sigs.k8s.io/structured-merge-diff/v4 v4.0.1/go.mod h1:bJZC9H9iH24zzfZ/41RGcq60oK1F7G282QMXDPYydCw=
 sigs.k8s.io/yaml v1.1.0/go.mod h1:UJmg0vDUVViEyp3mgSv9WPwZCDxu4rQW1olrI1uml+o=
+sigs.k8s.io/yaml v1.2.0 h1:kr/MCeFWJWTwyaHoR9c8EjH9OumOmoF9YGiZd7lFm/Q=
+sigs.k8s.io/yaml v1.2.0/go.mod h1:yfXDCHCao9+ENCvLSE62v9VSji2MKu5jeNfTrofGhJc=
```
Comment (yurkov-anton) [inline go.sum:840]: Is this file generated? How often it must be updated?

File: controllers/deployment_controller.go (hunk 5 above)
Comment (safronovD) [reply controllers/deployment_controller.go]: Oh, it is. Removed

File: go.sum (hunk 6 above)
Comment (safronovD) [reply go.sum:840]: It is autogenerated after go.mod updating

File: pkg/node/platform_description.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	labeltag      string
+	checkVersion
+}
+
+type checkVersion func(version string) bool
+
+var (
+	platforms = map[string]*PlatformDescription{
+		"default": {
+			imageName:     "csi-baremetal-node",
```
Comment (mishoyama) [inline pkg/node/platform_description.go:19]: Can `imageName` and `daemonsetName` be autogenerated?

File: pkg/node/platform_description.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+			imageName:     "csi-baremetal-node-kernel-5.4",
+			daemonsetName: "csi-baremetal-node-kernel-5.4",
+			labeltag:      "kernel-5.4",
+			checkVersion:  isNewKernel,
+		},
+	}
+)
+
+// isNewKernel returns true if kernel version >= 5.4
+func isNewKernel(version string) bool {
```
Comment (mishoyama) [inline pkg/node/platform_description.go:35]: `isNewKernel` name will be not relevant soon when we add `6.0` for example.

File: pkg/node/platform_description.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	kernelVersion := 5.4
```
Comment (mishoyama) [inline pkg/node/platform_description.go:36]: Please move to the const outside of the method.

File: controllers/deployment_controller.go [hunk 10]
```diff
@@ -106,23 +114,104 @@ func (r *DeploymentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
...
+	err = c.Watch(&source.Kind{Type: &appsv1.DaemonSet{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &csibaremetalv1.Deployment{},
+	})
+	if err != nil {
+		return err
+	}
+
+	// reconcile CSI Deployment if node was creates, node kernel-version or label were changed
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
```
Comment (mishoyama) [inline controllers/deployment_controller.go:150]: Can we have a filter to watch only those nodes which are labeled when deployed with node selector?

File: pkg/node/node.go [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+		if err != nil {
+			n.log.Error(err, "Failed to get kernel version for "+node.Name)
+			resultErr = err
+			continue
+		}
+
+		platformName := findPlatform(kernelVersion)
+		needToDeploy[platformName] = true
+
+		node.Labels[label] = platforms[platformName].labeltag
```
Comment (mishoyama) [inline pkg/node/node.go:131]: Do we need to remove this logic for node controller?

File: pkg/node/node.go [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+// selects fit platform for each one and add/update node platform-label
+// returns a Set of platforms, which will be deployed
+func (n *Node) updateNodeLabels() (Set, error) {
+	// need to trying getKernelVersion and update label on each node
+	// return err != nil to request reconcile again if one ore more nodes failed
+	var resultErr error
+
+	needToDeploy := createPlatformsSet()
+
+	nodes, err := n.clientset.CoreV1().Nodes().List(n.ctx, metav1.ListOptions{})
```
Comment (mishoyama) [inline pkg/node/node.go:115]: No need to mark all nodes only configured

**Review by mishoyama — CHANGES_REQUESTED** [review]

**Review by AnastasiaShemyakinskaya — APPROVED** [review]

File: pkg/node/platform_description.go (hunk 7 above)
Comment (safronovD) [reply pkg/node/platform_description.go]: Remove these fields from datastruct. Now names are chosen with functions

File: pkg/node/platform_description.go (hunk 8 above)
Comment (safronovD) [reply pkg/node/platform_description.go]: New signature `moreThan(version string, supported float64)`. Maybe when we support 6.0, we will need to add `between` or something else

File: pkg/node/platform_description.go (hunk 9 above)
Comment (safronovD) [reply pkg/node/platform_description.go]: fixed

File: controllers/deployment_controller.go (hunk 10 above)
Comment (safronovD) [reply controllers/deployment_controller.go]: Reworked. Currently reconcile request is returned only if watched node has an appropriate label from CSI Spec

File: pkg/node/node.go (hunk 11 above)
Comment (safronovD) [reply pkg/node/node.go:141]: Yes, we do

File: pkg/node/node.go (hunk 12 above)
Comment (safronovD) [reply pkg/node/node.go]: Fixed

**Review by mishoyama — APPROVED** [review]

**Pull request merged**